File: apps/frontend/src/api/endpoints/groceryLists.ts
import { apiClient } from '../client';

// Types for grocery list API
export interface GroceryListRequest {
  start_date: string; // ISO date string
  end_date: string; // ISO date string
}

export interface GroceryListIngredient {
  id: string;
  name: string;
  quantity_value: number;
  quantity_unit: string;
  recipes: string[];
}

export interface GroceryListResponse {
  start_date: string;
  end_date: string;
  ingredients: GroceryListIngredient[];
  total_meals: number;
}

export const groceryListsApi = {
  async generateGroceryList(
    request: GroceryListRequest
  ): Promise<GroceryListResponse> {
    return apiClient.request<GroceryListResponse>('/api/v1/grocery-lists', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  },
};

File: apps/backend/src/schemas/grocery_lists.py
"""Schemas for grocery list functionality."""

from __future__ import annotations

from datetime import date
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class GroceryListRequest(BaseModel):
    """Request model for generating a grocery list for a date range."""

    start_date: Annotated[
        date, Field(description="Start date for grocery list generation")
    ]
    end_date: Annotated[date, Field(description="End date for grocery list generation")]

    model_config = ConfigDict(extra="forbid")


class GroceryListIngredient(BaseModel):
    """Individual ingredient in a grocery list with aggregated quantity."""

    id: Annotated[UUID, Field(description="Ingredient identifier")]
    name: Annotated[str, Field(description="Ingredient name")]
    quantity_value: Annotated[
        float, Field(description="Total aggregated quantity across all recipes")
    ]
    quantity_unit: Annotated[str, Field(description="Unit for the quantity")]
    recipes: Annotated[
        list[str], Field(description="Names of recipes that use this ingredient")
    ]

    model_config = ConfigDict(extra="forbid")


class GroceryListResponse(BaseModel):
    """Response model for grocery list containing aggregated ingredients."""

    start_date: Annotated[date, Field(description="Start date of the grocery list")]
    end_date: Annotated[date, Field(description="End date of the grocery list")]
    ingredients: Annotated[
        list[GroceryListIngredient], Field(description="List of aggregated ingredients")
    ]
    total_meals: Annotated[
        int, Field(description="Total number of meals included in the date range")
    ]

    model_config = ConfigDict(extra="forbid")
